pkg/docker/config/config.go

```diff
@@ -0,0 +1,276 @@
+package config
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	helperclient "github.com/docker/docker-credential-helpers/client"
+	"github.com/docker/docker-credential-helpers/credentials"
+	"github.com/docker/docker/pkg/homedir"
+	"github.com/pkg/errors"
+)
+
+type dockerAuthConfig struct {
+	Auth string `json:"auth,omitempty"`
+}
+
+type dockerConfigFile struct {
+	AuthConfigs map[string]dockerAuthConfig `json:"auths"`
+	CredHelpers map[string]string           `json:"credHelpers,omitempty"`
+}
+
+const (
+	dockerHostname = "docker.io"
+	dockerRegistry = "registry-1.docker.io"
+
+	authCfg           = ".config/containers"
+	authCfgFileName   = "auth.json"
+	dockerCfg         = ".docker"
+	dockerCfgFileName = "config.json"
+)
+
+var (
+	// ErrNotLoggedIn is returned for users not logged into a registry
+	// that they are trying to logout of
+	ErrNotLoggedIn = errors.Errorf("not logged in")
+)
+
+// SetAuthentication stores the username and password in the auth.json file
+func SetAuthentication(registry, username, password string) error {
+	registry = dockerHostNameToRegistry(registry)
+	return modifyJSON(func(auths *dockerConfigFile) (bool, error) {
+		if len(auths.CredHelpers) > 0 {
+			if err := setAuthToCredHelper(registry, username, password); err != nil {
+				return false, err
+			}
+			return true, nil
```

Comment:
This should return `false, nil` AFAICT; IOW `return false, setAuthToCredHelper(…)`.

Same in `RemoveAuthentication`.

Reply:
If `setAuthToCredHelper()` succeeds we want to return true, otherwise `modifyJSON` will not know that the dockerConfigFile has been updated.

Reply:
But it hasn’t been updated; we have just read the helper identity, and called the helper.

Reply:
The helper is returning `helperclient.Store()`, which I thought adds the information to the credHelper?

Reply:
That executes a binary *-`credHelper` (, which stores it _somewhere_); it does not store it to the `auths` data structure which the code is writing back (the helper doesn’t receive any reference to  that structure, after all).

Reply:
oh okay, got it! thanks :)